### PR TITLE
Fix some minor issues with references.

### PR DESCRIPTION
**Pull request by AndrolGenhald**
References assigned to properties on `$this` should never be unused.
Using a `@var` docblock before a reference should be allowed if it targets a variable instead of the assignment statement.

### CODE REVIEW

File: src/Psalm/Internal/Analyzer/Statements/Expression/AssignmentAnalyzer.php [hunk 1]
```diff
@@ -1004,6 +1005,15 @@ public static function analyzeAssignmentRef(
                         : "property-assignment-as-reference"
                 );
             }
+
+            if ($root_var_id === '$this') {
+                // Variables on `$this` are always used
```
[the remark this answers is not in the document]
Comment (AndrolGenhald) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssignmentAnalyzer.php:1010]: No, it constitutes use of `$this->whatever` as well. We don't currently seem to track [unused properties](https://psalm.dev/r/9176b3c10d), and [this](https://psalm.dev/r/baa101e336) caused me some issues.

[the remark this answers is not in the document]
Comment (AndrolGenhald) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssignmentAnalyzer.php:1010]: [Reference assignment](https://psalm.dev/r/c9b6142e68) actually catches more than [regular assignment](https://psalm.dev/r/c8d4aca0b4) for properties. I'm not too sure if it'd be better to just make it work the same or if non-reference assignment should be improved, so I decided to start small and make `$this` a special case.

[the remark this answers is not in the document]
Comment (weirdan) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssignmentAnalyzer.php:1010]: > We don't currently seem to track [unused properties](https://psalm.dev/r/9176b3c10d),

We actually do, but you have to use the class itself and enable `Detect unused classes and methods` checkbox on psalm.dev: https://psalm.dev/r/031e79174f

[the remark this answers is not in the document]
Comment (AndrolGenhald) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssignmentAnalyzer.php:1010]: Ah, I missed that the class had to be used.

[the remark this answers is not in the document]
Comment (orklah) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssignmentAnalyzer.php:1010]: Seems ok if it only involves references

**Comment by orklah** [comment]
Thanks!

**Pull request merged**